From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Olexa Bilaniuk <obilaniu@gmail.com>
Date: Wed, 23 Feb 2022 13:54:01 -0500
Subject: [PATCH] Enable renaming main functions


diff --git a/lua.c b/lua.c
index 7f7dc2b2..28abd014 100644
--- a/lua.c
+++ b/lua.c
@@ -658,7 +658,10 @@ static int pmain (lua_State *L) {
 }
 
 
-int main (int argc, char **argv) {
+#ifndef main_lua
+#define main_lua main
+#endif
+int main_lua (int argc, char **argv) {
   int status, result;
   lua_State *L = luaL_newstate();  /* create state */
   if (L == NULL) {
diff --git a/luac.c b/luac.c
index 52f41ca0..4dd3bf41 100644
--- a/luac.c
+++ b/luac.c
@@ -196,7 +196,10 @@ static int pmain(lua_State* L)
  return 0;
 }
 
-int main(int argc, char* argv[])
+#ifndef  main_luac
+# define main_luac main
+#endif
+int main_luac(int argc, char* argv[])
 {
  lua_State* L;
  int i=doargs(argc,argv);
